gapic/src/main/com/google/gapid/perfetto/models/Tracks.java

```diff
@@ -157,11 +174,38 @@ private Tracks() {
             group(state -> new TitlePanel("GPU Counters"), true));
         parent = "gpu_counters";
       }
-      for (CounterInfo counter : counters) {
-        CounterTrack track = new CounterTrack(data.qe, counter);
-        data.tracks.addTrack(parent, track.getId(), counter.name,
-            single(state -> new CounterPanel(state, track, DEFAULT_COUNTER_TRACK_HEIGHT), true,
-                /*right truncate*/ true));
+      Multimap<Long, Long> groups = data.getGpuCounterGroups();
+      if (groups.keySet().size() > 1) {
+        for (GpuCounterGroup group : GpuCounterGroup.values()) {
+          if (group == GpuCounterGroup.UNRECOGNIZED) {
+            continue;
+          }
+          Collection<Long> grouped_counters = groups.get(Long.valueOf(group.getNumber()));
+          if (grouped_counters.isEmpty()) {
+            continue;
+          }
+          parent = "gpu_counters_group_" + group.name();
+          String name = GPU_COUNTER_GROUP_NAMES.get(group);
+          data.tracks.addLabelGroup("gpu_counters", parent, name,
+              group(state -> new TitlePanel(name), true));
+          for (Long counter_id : grouped_counters) {
+            CounterInfo counter = counters.get(counter_id);
+            if (counter == null) {
+              continue;
+            }
+            CounterTrack track = new CounterTrack(data.qe, counter);
```

Comment:
I just realized that the way it's written now, a counter can be in more than one group. You can only instantiate a `CounterTrack` for each counter once. If we want a counter to show up multiple times, we can have multiple `CounterPanel` instances, but they all need to share the same `CounterTrack` instance.

Reply:
Done